src/components/Layout/TableOfContents.jsx

```diff
@@ -1,165 +1,90 @@
-  constructor() {
-    super();
-    this.nodeListItemsToRender = [];
-    this.currentLevel = 0;
-  }
-
-  formatChapterTitle(title) {
-    return title.split('_').map(word => word[0].toUpperCase() + word.substring(1)).join(' ');
-  }
-
-  addSubchapterJSX(node) {
-    this.nodeListItemsToRender.push(
-      <SubchapterLIContainer key={node}>
-        <h5>
-          {this.formatChapterTitle(node)}
-        </h5>
-      </SubchapterLIContainer>
-    );
-  }
-
-  addChapterJSX(node) {
-    this.nodeListItemsToRender.push(
-      <ChapterLIContainer key={node}>
-        <h5>
-          {this.formatChapterTitle(node)}
-        </h5>
-      </ChapterLIContainer>
-    )
-  }
-
-  buildLessonItemNodes(nodeArray) {
-    nodeArray.forEach(node => {
-      this.nodeListItemsToRender.push(
-        <LessonLIContainer key={node.post.id}>
-          <Link to={node.post.childMarkdownRemark.fields.slug}>
-            <li>
-              <h6>{node.post.childMarkdownRemark.frontmatter.title}</h6>
-            </li>
-          </Link>
-        </LessonLIContainer>
-      )
-    })
-  }
-
-  buildChapterNodes(node) {
-    // If this is a Chapter (and not a subchapter)
-    if (this.currentLevel === 0) {
-      this.addChapterJSX(node);
-      // Else it's a SubChapter
-    } else {
-      this.addSubchapterJSX(node);
+import React from "react";
+import Link from "gatsby-link";
+import styled from "styled-components";
+
+const Links = ({ entries, index }) => (
+  <StyledLinkList>
+    {entries.map(({ entry }, key) => (
+      <EntryListItem key={`link${index}${key}`}>
+        <Link to={entry.childMarkdownRemark.fields.slug}>
+          <EntryTitle>{entry.childMarkdownRemark.frontmatter.title}</EntryTitle>
+        </Link>
+      </EntryListItem>
+    ))}
+  </StyledLinkList>
+);
+
+const Subchapters = ({ chapters, index }) => (
+  <StyledSubchapterList>
+    {chapters.map(({ entries, title }, key) => (
+      <li key={`subchapter${index}${key}`}>
+        <SubchapterTitle>{title}</SubchapterTitle>
+        <Links entries={entries} index={`subchapter${index}${key}`} />
+      </li>
+    ))}
+  </StyledSubchapterList>
+);
+
+const chapterMapper = ({ subchapters, entries, title }, index) => {
+  const AdditionalElements = () => {
+    if (subchapters) {
+      return <Subchapters chapters={subchapters} index={index} />;
     }
-  }
-
-
-  // Level matters because sub-chapter <li> and chapter <li>s are styled differently
-  //
-  // If the node is an Array, it holds the actual page nodes itself,
-  // 1. Add the node's value as either a chapter <li> or subchapter <li> - based on level
-  // 2. Build the lesson <li> links
-  //
-  // Else, its an object, so it must be a chapter or sub-chapter
-  // 1. build the Chapter <li>
-  // 2. Recursively restart on the child node
-  //
-  // For level, if we've hit an array, then we must be as deeply nested as possible.
-  // So the next node we hit will be a top level chapter node, so we're back to level 0.
-  //
```

Comment:
I admit that my approach does not (in this simple version) allow for arbitrary depth of the Chapter - Subchapter - ... construct. But I also think that it does not make sense to support this in the boilerplate. Also it is easy to adjust my approach with a little recursion to allow for any kind of nesting some sick mind could come up with.

Reply:
I agree with every comment in here, except this one. My reasoning for redoing the ToC in the way that it is now is because I needed the freedom to change the order and nesting of ToC without much work.

I agree that repeating the structure in the GraphQL is an issue. It's getting cumbersome as my site grows (the one I built this for). It also leaves a lot of room for silly typo errors.

Can we think of a way that allows for an arbitrary number of nested levels, but maintain all the benefits that your changes would provide?